pkg/kv/kvserver/rangefeed/buffered_registration.go

```diff
@@ -196,16 +209,29 @@ func (br *bufferedRegistration) outputLoop(ctx context.Context) error {
 			overflowed = br.mu.overflowed
 			br.mu.caughtUp = true
 		}
+		if firstIteration {
+			wasOverflowedOnFirstIteration = br.mu.overflowed
+		}
 		br.mu.Unlock()
+		firstIteration = false
+
+		if wasOverflowedOnFirstIteration && br.shouldLogOverflow(oneCheckpointWithTimestampSent) {
+			log.Warningf(ctx, "rangefeed %s overflowed during catch up scan from %s (useful checkpoint sent: %v)",
+				br.span, br.catchUpTimestamp, oneCheckpointWithTimestampSent)
```

Comment:
I might be missing something here, but could you help me understand the reason behind logging this message multiple times. Since `oneCheckpointWithTimestampSent` is a boolean, will the message be the same every time? It seems useful if we could log the exact checkpoint events struct we sent previously rather than the boolean.

Reply:
If `wasOverflowedOnFirstIteration` was true then `overflowed` is necessarily true. So, for a given run of this loop, we won't run this multiple times because we will return at 223 below.

> It seems useful if we could log the exact checkpoint events struct we sent previously rather than the boolean.

For the condition we care about most here (overflowed without a useful checkpoint), there is no additional info in the checkpoint. It'll have the span that is already logged and an empty timestamp.

Reply:
You are right, I got confused. Thanks!

Reply:
> If wasOverflowedOnFirstIteration was true then overflowed is necessarily true. 

I'm getting a bit confused, so I might be missing something obvious. wasOverflowedOnFirstIteration is set to true if `br.mu.overflowed` is true. https://github.com/cockroachdb/cockroach/blob/79cb51a754f83001ffda5eb5591d7174e4bc4053/pkg/kv/kvserver/rangefeed/buffered_registration.go#L213 But we don't set `overflowed ` to true until we empty the buffer https://github.com/cockroachdb/cockroach/blob/79cb51a754f83001ffda5eb5591d7174e4bc4053/pkg/kv/kvserver/rangefeed/buffered_registration.go#L208-L209. So we would not hit this return at 223 until we empty the buffer. Does that seem right?

Reply:
Ah! Thank you, I moved some code around on rebase and definitely broke this.

Reply:
Just put it back to using the nested ifs. Apologies for not looking more closely the first time.